src/app/EventManagement.cpp

```diff
@@ -0,0 +1,903 @@
+
+    VerifyOrExit(mState != EventManagementStates::Shutdown, err = CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrExit(apOptions != nullptr && apOptions->mpEventSchema != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
+
+    err = LogEventPrivate(apDelegate, apOptions, aEventNumber);
+
+exit:
+    Platform::CriticalSectionExit();
+    ChipLogFunctError(err);
+    return err;
+}
+
+CHIP_ERROR EventManagement::LogEventPrivate(EventLoggingDelegate * apDelegate, const EventOptions * apOptions,
+                                            chip::EventNumber & aEventNumber)
+{
+    CircularTLVWriter writer;
+    CHIP_ERROR err                 = CHIP_NO_ERROR;
+    size_t requestSize             = kEventSizeRezerve;
+    bool didWriteEvent             = false;
+    aEventNumber                   = 0;
+    CircularEventBuffer checkpoint = *mpEventBuffer;
+
+    EventLoadOutContext ctxt = EventLoadOutContext(writer, apOptions->mpEventSchema->mPriority,
+                                                   GetPriorityBuffer(apOptions->mpEventSchema->mPriority)->mLastEventNumber);
+    Timestamp timestamp(Timestamp::Type::kSystem, System::Timer::GetCurrentEpoch());
+    EventOptions opts = EventOptions(timestamp);
+
+    // check whether the entry is to be logged or discarded silently
+    VerifyOrExit(apOptions->mpEventSchema->mPriority >= CHIP_CONFIG_EVENT_GLOBAL_PRIORITY, /* no-op */);
+
+    // Create all event specific data
+    // Timestamp; encoded as a delta time
+    if ((apOptions->mTimestamp.mType == Timestamp::Type::kSystem))
+    {
+        opts.mTimestamp = apOptions->mTimestamp;
+    }
+
+    if (GetPriorityBuffer(apOptions->mpEventSchema->mPriority)->mFirstEventSystemTimestamp.mValue == 0)
+    {
+        GetPriorityBuffer(apOptions->mpEventSchema->mPriority)->AddEvent(opts.mTimestamp);
+    }
+
+    opts.mUrgent       = apOptions->mUrgent;
+    opts.mpEventSchema = apOptions->mpEventSchema;
+
+    ctxt.mFirst              = false;
+    ctxt.mCurrentEventNumber = GetPriorityBuffer(opts.mpEventSchema->mPriority)->mLastEventNumber;
+    ctxt.mCurrentSystemTime  = GetPriorityBuffer(opts.mpEventSchema->mPriority)->mLastEventSystemTimestamp;
+
+    // Begin writing
+    while (!didWriteEvent)
+    {
+        // Ensure we have space in the in-memory logging queues
+        err = EnsureSpace(requestSize);
+        // If we fail to ensure the initial reserve size, then the
+        // logging subsystem will never be able to make progress.  In
+        // that case, it is best to assert
+        if ((requestSize == kEventSizeRezerve) && (err != CHIP_NO_ERROR))
+            chipDie();
+        SuccessOrExit(err);
+
+        // save a checkpoint for the underlying buffer.  Note that with
+        // the current event buffering scheme, only the mpEventBuffer will
+        // be affected by the writes to the `writer` below, and thus
+        // that's the only thing we need to checkpoint.
+        checkpoint = *mpEventBuffer;
```

Comment:
We have quite a bit of these copies and this is a common operation. I wonder if it is worth it us copying over things like this. How about defining a separate type of 'Snapshot' that only includes relevant bits and then the code is clearer:

```cpp
Snapshot checkpint = mpEventBufferf->CreateCheckpoint();

....

mpEventBuffer->RestoreCheckpoint(checkpoint);

```

seems easier to read than operator= and would allow us to make the checkpoint potentially smaller by not including bits that are not relevant (are some pointers constant, are the counters constant?)

Reply:
after optimizing the code, only one mpEventBuffer's checkpoint exists, it seems there is no necessity to modulize this logic, thanks